### ISSUE TITLE
transfert block GMAIL. "Invalid character in literal"

### ISSUE THREAD

**Issue by omlet05** [imapsync/imapsync#124]
Hello,

I have problem during sync with big folder: about 35000. My sync is from Zimbra to GMAIL.

- msg Sent Messages/209988 {2241470} couldn't append  (Subject:[]) to folder [Gmail]/Sent Mail: * BYE Invalid character in literal.

The sync always get closed by this mail. Any way to bypass it or delete it easily?

**Comment by gilleslamiral** [comment]
You can avoid this message using `--search1 "NOT UID 209988"` but a better way would be to understand why Gmail complain about and quit with the message "Invalid character in literal.".

--debugcontent will print the message, --debuimap2 will debug the imap session but with a 2 MB message it is an hard work to find and fix the issue.

**Comment by omlet05** [comment]
Thanks a lot, I'll try to bypass this mail at this time the migration is urgent...


I'm going to check why it blocks after.

**Comment by pgera** [comment]
I see this error too. I enabled --debugcontent and --debuimap2, and have the logs. I don't see which character it tripped on though in the logs. The message itself is a short message (13774 bytes). Where do I go from here ? The error appears to come from /usr/share/perl5/vendor_perl/Mail/IMAPClient.pm line 1415. This is on a RHEL 7 system (with Perl 5.16.3 and Mail::IMAPClient 3.37). I can share the logs privately if it helps.

**Comment by pgera** [comment]
reformime didn't help. I also checked if any character in the message > 127, but there wasn't any.

**Comment by pgera** [comment]
Sent you an email

**Comment by gilleslamiral** [comment]
Thanks for the log. There is a NUL character (NUL == 0) at the end of the buggy messages.
Imapsync to the rescue:
```
# remove all NUL characters with:

    imapsync ...  --regexmess  's/\000//g'
```
Usage:
```
     --regexmess    reg  : Apply the whole regex to each message before transfer.
                           Example: 's/\000/ /g' # to replace null by space.
     --regexmess    reg  : and this one, etc.
```
Tell me if I'm right, then I may add the regex for  the option` --gmail2`

**Comment by pgera** [comment]
Indeed, that solves the error. Thanks a lot. After some cursory googling, I found this: https://stackoverflow.com/a/12879473

> Old email (according to RFC 822) was allowed to have null characters. New email (according to RFC 2822, circa 2003) is not allowed to have null characters. Note RFC 2822 reads: "Differences from earlier standards... ASCII 0 (null) removed."

>It's entirely possible that Gmail accepts 822-style emails via SMTP (that's how the email first got to your inbox) but only 2822-style emails via IMAP (which is why you can't put it back via IMAP).

>** Note: Don't blindly strip nulls from MIME documents included in the email. RFC 2822 "specifies that messages are made up of characters in the US-ASCII range of 1 through 127. There are other documents, specifically the MIME document series [RFC2045, RFC2046, RFC2047, RFC2048, RFC2049], that extend [RFC 2822] to allow for values outside of that range."

**Comment by gilleslamiral** [comment]
Good!

The comment "ASCII 0 (null) removed." in rfc2822 is for the header part of a message.

[rfc2822#section-4](https://tools.ietf.org/html/rfc2822#section-4) 

>    Finally, certain characters that were formerly allowed in messages
>    appear in this section.  The NUL character (ASCII value 0) was once
>    allowed, but is no longer for compatibility reasons.  CR and LF were
>    allowed to appear in messages other than as CRLF; this use is also
>    shown here.

So yes, some messages can have 8bit in the body, there is a mime for this I remember.

**Comment by pgera** [comment]
Does the regexmess argument apply to the entire message, or just the header ? We need this only for the header, right ?

**Comment by gilleslamiral** [comment]
Problem solved. Closing.

**Comment by nickcobley** [comment]
I have been running into this issue quite a lot with gmail to gmail migrations, I have been hitting it very frequently in very large migrations of mailboxes with over 200k emails. Today it got to a point where the we were talking hundreds of emails in a mailbox were failing with this error.

Adding the `--regexmess  's/\000//g'` fix from above fixed it right away. I'm going to keep trying this on new mailboxes, certainly seems like it would be a good candidate to add to --gmail2 though.

**Comment by mosteo** [comment]
Same here, I encountered this issue with a gmail to gmail migration. In my case 4 emails of ~200k.
